### PR TITLE
new utilities for reshaping and predetermining the shapes of n-d DWT coefficients

### PR DESCRIPTION
**Pull request by grlee77**
The overall point of this PR is to introduce a few convenience routines for working with n-dimensional DWTs via `wavedecn` and `waverecn`.  I have found need for this type of thing in my own work involving wavelet thresolding as a form of regularization during iterative algorithms.

As a concrete example, the [operator discreization library](https://github.com/odlgroup/odl) currently has a wavelet transform operator based on PyWavelets.  The backend code for this currently involves somewhat complicated bookkeeping and has become out of sync with some newer features of `wavedecn` (unique wavelet or boundary mode can be defined per axis).  With the addition of a few minor utility functions within PyWavelets we could remove the need for almost all of the backend code found here:
https://github.com/odlgroup/odl/blob/master/odl/trafos/backends/pywt_bindings.py

In this PR I add three primary things:

1.) `ravel_coeffs` (and `unravel_coeffs`):  This takes the output of `wavedec`, `wavedec2` or `wavedecn` and stacks all coefficients into a single contiguous 1D array.  The inverse operation is done by `unravel_coeffs`.  In some specific cases this can already easily be done by just ravelling the output of the existing PyWavelets functions `coeffs_to_array` and `array_to_coeffs`.  The primary difference is that for some boundary-modes, `coeffs_to_array` does not produced a tightly packed n-d array so that the total array size is not always identical to the number of wavelet coefficients.

2.)  `dwtn_max_level` is an n-dimensional generalization of the existing `dwt_max_level`.  It just applies `dwt_max_level` along each axis to be transformed and then takes the minimum across the axes.  This allows an easy way to compute the maximum feasible level when calling `wavedecn`.

3.) `wavedecn_shapes` and `wavedecn_size` are new routines that compute the total number of coefficients or the shapes of all coefficients that are going to be produced by an n-dimensional transform.  The only way to easily do this currently is to actually first run a transform and then extract the shapes from the coefficients produced.  This is useful in the context of creating operators as in `odl` because one must specify appropriate shapes for the transform domain at the time of operator creation.

cc: @adler-j and @kohr-h as odl devs who may be interested and/or have feedback as well.  I can open a separate PR over at `odl` demonstrating how the code there could be simplified using these proposed routines.

TODO:
- [x] add tests for `dwtn_max_level` and `wavedecn_shapes`

### CODE REVIEW

**Comment by kohr-h** [comment]
@grlee77 👍 This would be super useful!

**Comment by rgommers** [comment]
Looks like a good idea. Still needs tests.

**Comment by grlee77** [comment]
This should be ready for final review. I extended the test coverage and added the new functions to the API documentation.

A concrete demonstration of how this PR can be used to greatly simplify ODL pywt backend code (while also adding adding `axes` support to the `WaveletTransform` operator):
https://github.com/odlgroup/odl/compare/odlgroup:9020fd5...grlee77:5a6d373

File: pywt/_multilevel.py [hunk 1]
```diff
@@ -826,3 +853,296 @@ def array_to_coeffs(arr, coeff_slices, output_format='wavedecn'):
...
+        Maximum level.
+
+    Notes
+    -----
+    The level returned is the smallest `dwt_max_level` over all axes.
+
+    Examples
+    --------
+    >>> import pywt
+    >>> max_level = pywt.dwtn_max_level((64, 32), 'db2')
```
Comment (rgommers) [inline pywt/_multilevel.py]: remove `max_level = ` and show output `3`

File: pywt/_multilevel.py [hunk 2]
```diff
@@ -826,3 +853,296 @@ def array_to_coeffs(arr, coeff_slices, output_format='wavedecn'):
...
+
+
+def ravel_coeffs(coeffs, axes=None):
+    """Ravel a set of multilevel wavelet coefficients into a single 1D array.
+
+    Parameters
+    ----------
+    coeffs : array-like
+        A list of multilevel wavelet coefficients as returned by
+        ``wavedec``, ``wavedec2`` or ``wavedecn``.
```
Comment (rgommers) [inline pywt/_multilevel.py]: single rather than double backticks

File: pywt/_multilevel.py [hunk 3]
```diff
@@ -826,3 +853,296 @@ def array_to_coeffs(arr, coeff_slices, output_format='wavedecn'):
...
+        default value of ``None`` corresponds to all axes.
+
+    Returns
+    -------
+    coeff_arr : array-like
+        Wavelet transform coefficient array. All coefficients have been
+        concatenated into a single array.
+    coeff_slices : list
+        List of slices corresponding to each coefficient. As a 2D example,
+        `coeff_arr[coeff_slices[1]['dd']]` would extract the first level detail
```
Comment (rgommers) [inline pywt/_multilevel.py]: double backticks here for typesetting as code

File: pywt/_multilevel.py [hunk 4]
```diff
@@ -826,3 +853,296 @@ def array_to_coeffs(arr, coeff_slices, output_format='wavedecn'):
...
+        coefficients from `coeff_arr`.
+    coeff_shapes : list
+        List of shapes corresponding to each coefficient. For example, in 2D,
+        coeff_shapes[1]['dd'] would contain the original shape of the first
```
Comment (rgommers) [inline pywt/_multilevel.py]: double backticks around this code snippet

File: pywt/_multilevel.py [hunk 5]
```diff
@@ -826,3 +853,296 @@ def array_to_coeffs(arr, coeff_slices, output_format='wavedecn'):
...
+            coeff_shapes[-1][key] = d.shape
+    return coeff_arr, coeff_slices, coeff_shapes
+
+
+def unravel_coeffs(arr, coeff_slices, coeff_shapes, output_format='wavedecn'):
+    """Unravel a raveled array of multilevel wavelet coefficients.
+
+    Parameters
+    ----------
+
```
Comment (rgommers) [inline pywt/_multilevel.py]: spurious blank line

File: pywt/_multilevel.py [hunk 6]
```diff
@@ -826,3 +853,296 @@ def array_to_coeffs(arr, coeff_slices, output_format='wavedecn'):
...
+    arr : array-like
+        An array containing all wavelet coefficients. This should have been
+        generated by applying``ravel_coeffs`` to the output of ``wavedec``,
```
Comment (rgommers) [inline pywt/_multilevel.py]: single backticks, and a typo (missing space after applying)

File: pywt/_multilevel.py [hunk 7]
```diff
@@ -826,3 +853,296 @@ def array_to_coeffs(arr, coeff_slices, output_format='wavedecn'):
...
+        ``wavedec2`` or ``wavedecn``.
+    coeff_slices : list of tuples
+        List of slices corresponding to each coefficient as obtained from
+        ``ravel_coeffs``.
+    coeff_shapes : list of tuples
+        List of shapes corresponding to each coefficient as obtained from
+        ``ravel_coeffs``.
+    output_format : {'wavedec', 'wavedec2', 'wavedecn'}
```
Comment (rgommers) [inline pywt/_multilevel.py]: add `, optional` and "default is wavedecn"

**Comment by grlee77** [comment]
Thanks. I agree with the comments and have addressed them in the first commit here. The last one made similar changes throughout the rest of the file. NumPy's docstring format documentation doesn't seem to put None in backticks so I did the same here.

One other question: If we want crosslinks to functions appear in the docstrings, don't we need to use the form:
:func:`pywt.wavedec`, :func:`pywt.wavedec2` or :func:`pywt.wavedecn`? I tried locally in `ravel_coeffs` and doing it that way results in links to these other functions rather than just displaying the names in italics.

**Comment by rgommers** [comment]
`:func:` should not be necessary. that'd be a bit ugly, the plain-text version becomes harder to read.

if that doesn't work right now then there's something on our Sphinx conf related to numpydoc not set up correctly. I'll have a look.

**Comment by rgommers** [comment]
Okay, in it goes! Thanks Greg.

**Pull request merged**